internal/oidc/clientregistry/clientregistry.go

```diff
@@ -11,47 +11,109 @@ import (
 type Client struct {
 	fosite.DefaultOpenIDConnectClient
 }
 
-func (c Client) GetResponseModes() []fosite.ResponseModeType {
-	// For now, all Pinniped clients always support "" (unspecified), "query", and "form_post" response modes.
-	return []fosite.ResponseModeType{fosite.ResponseModeDefault, fosite.ResponseModeQuery, fosite.ResponseModeFormPost}
-}
-
-// It implements both the base, OIDC, and response_mode client interfaces of Fosite.
+// Client implements the base, OIDC, and response_mode client interfaces of Fosite.
 var (
 	_ fosite.Client              = (*Client)(nil)
 	_ fosite.OpenIDConnectClient = (*Client)(nil)
 	_ fosite.ResponseModeClient  = (*Client)(nil)
 )
 
-// StaticClientManager is a fosite.ClientManager with statically-defined clients.
-type StaticClientManager struct{}
+func (c Client) GetResponseModes() []fosite.ResponseModeType {
+	if c.ID == PinnipedCLIClientID {
+		// The pinniped-cli client supports "" (unspecified), "query", and "form_post" response modes.
+		return []fosite.ResponseModeType{fosite.ResponseModeDefault, fosite.ResponseModeQuery, fosite.ResponseModeFormPost}
+	}
+	// For now, all other clients support only "" (unspecified) and "query" response modes.
+	return []fosite.ResponseModeType{fosite.ResponseModeDefault, fosite.ResponseModeQuery}
+}
+
+// ClientManager is a fosite.ClientManager with a statically-defined client and with dynamically-defined clients.
+type ClientManager struct {
+	oidcClientsClient supervisorclient.OIDCClientInterface
+	storage           *oidcclientsecretstorage.OIDCClientSecretStorage
+	minBcryptCost     int
+}
 
-var _ fosite.ClientManager = (*StaticClientManager)(nil)
+var _ fosite.ClientManager = (*ClientManager)(nil)
 
-// GetClient returns a static client specified by the given ID.
+func NewClientManager(
+	oidcClientsClient supervisorclient.OIDCClientInterface,
+	storage *oidcclientsecretstorage.OIDCClientSecretStorage,
+	minBcryptCost int,
+) *ClientManager {
+	return &ClientManager{
+		oidcClientsClient: oidcClientsClient,
+		storage:           storage,
+		minBcryptCost:     minBcryptCost,
+	}
+}
+
+// GetClient returns the client specified by the given ID.
 //
 // It returns a fosite.ErrNotFound if an unknown client is specified.
-func (StaticClientManager) GetClient(_ context.Context, id string) (fosite.Client, error) {
-	switch id {
-	case "pinniped-cli":
+// Other errors returned are plain errors, because fosite will wrap them into a new ErrInvalidClient error and
+// use the plain error's text as that error's debug message (see client_authentication.go in fosite).
+func (m *ClientManager) GetClient(ctx context.Context, id string) (fosite.Client, error) {
+	if id == PinnipedCLIClientID {
+		// Return the static client. No lookups needed.
 		return PinnipedCLI(), nil
-	default:
+	}
+
+	// Try to look up an OIDCClient with the given client ID (which will be the Name of the OIDCClient).
+	oidcClient, err := m.oidcClientsClient.Get(ctx, id, v1.GetOptions{})
```

Comment:
Before doing the lookup, I would check if the name has the required prefix.

Reply:
Updated in 88f611d31a07382565ca78485980023ac8ef2145.